### PR TITLE
Fix Windows.Applications.Chrome.History

### LINKED ISSUE #1538: Windows.Applications.Chrome.History only shows most recent visit
During artefact testing and validation, I found that this artefact only returns the most recent visit to a given URL.

image
image

Looking in the SQLite database these queries can potentially be fixed by joining the _url_ table to the _visits_ table.

I am currently working on a fix for Pull Request.

### PR DESCRIPTION
**Pull request by angry-bender**
Fixes  #1538

Tested with chrome v 98.0.4758.80 and Windows 10

Verification screenshot
image

### CODE REVIEW

**Comment by mgreen27** [comment]
Good job!

I wonder if its worth consolidating to a general Browser history (its just really adding to the glob)? 

(Credit the SQL below to one of my colleagues who also did similar join previously)

```
name: Windows.Applications.Browser.History
description: |
  Enumerate the users chrome and edge history. specifically the urls table and will join based on values from the visits table.

parameters:
  - name: KeyGlobs
    default: '\AppData\Local\{Google\Chrome,Microsoft\Edge}\User Data\*\History'
  - name: urlSQLQuery
    default: |
      SELECT visits.url as visit_url_id, visits.visit_time as url_visit_time, urls.id as url_id, urls.url as visited_url, urls.title as url_title, urls.visit_count as url_visit_count,
             urls.typed_count as urls_typed_count, urls.last_visit_time as url_last_visit_time
      FROM urls
      INNER JOIN visits ON visits.url=urls.id
  - name: UserRegex
    default: .
    
export: |
     LET browser_history(path,sql_query) =SELECT 
            User, 
            FullPath,
            Mtime,
            visit_url_id,
            timestamp(winfiletime=url_visit_time * 10) as url_visit_time,
            url_id,
            visited_url,
            url_title,
            url_visit_count,
            urls_typed_count,
            timestamp(winfiletime=url_last_visit_time * 10) as url_last_visit_time
        FROM sqlite(
            file=path,
            query=sql_query)
          
          
sources:
  - query: |
      LET find_users = SELECT Uid, Name AS User, Directory
        FROM Artifact.Windows.Sys.Users()
        WHERE Name =~ UserRegex
        
      LET history_files = SELECT * FROM foreach(
          row=find_users,
          query={
             SELECT User, FullPath, Mtime FROM glob(globs=Directory + KeyGlobs)
          })
          
      SELECT * FROM foreach(row=history_files,
            query={ SELECT * FROM browser_history(path=FullPath,sql_query=urlSQLQuery) 
        })
```

File: artifacts/definitions/Windows/Applications/Chrome/History.yaml [hunk 1]
```diff
@@ -31,11 +33,16 @@ sources:
...
         SELECT * FROM foreach(row=history_files,
           query={
-            SELECT User, FullPath,
-                   Mtime,
-                   visited_url,
-                   title, visit_count, typed_count,
-                   timestamp(winfiletime=last_visit_time * 10) as last_visit_time
+            SELECT User, 
+                   id AS url_id,
+                   timestamp(epoch=((visit_time / 1000000) - 11644473600)) AS visit_time,
```
Comment (scudette) [inline artifacts/definitions/Windows/Applications/Chrome/History.yaml:38]: you can just do `timestamp(winfiletime=visit_time)`

Comment (angry-bender) [reply artifacts/definitions/Windows/Applications/Chrome/History.yaml:38]: Oh thats awesome. I'll test that now

Comment (angry-bender) [reply artifacts/definitions/Windows/Applications/Chrome/History.yaml]: Thats worked a treat, thanks for that. I didn't realise a chrome timestamp was the same as that. Fixed in [ff7e381](https://github.com/angry-bender/velociraptor/commit/ff7e38115356d6c5f18a80171c47e120c823d3f6)

image

**Comment by angry-bender** [comment]
> Good job!
> 
> I wonder if its worth consolidating to a general Browser history (its just really adding to the glob)?
> 
> (Credit the SQL below to one of my colleagues who also did similar join previously)
> 
> ```
> name: Windows.Applications.Browser.History
> description: |
>   Enumerate the users chrome and edge history. specifically the urls table and will join based on values from the visits table.
> 
> parameters:
>   - name: KeyGlobs
>     default: '\AppData\Local\{Google\Chrome,Microsoft\Edge}\User Data\*\History'
>   - name: urlSQLQuery
>     default: |
>       SELECT visits.url as visit_url_id, visits.visit_time as url_visit_time, urls.id as url_id, urls.url as visited_url, urls.title as url_title, urls.visit_count as url_visit_count,
>              urls.typed_count as urls_typed_count, urls.last_visit_time as url_last_visit_time
>       FROM urls
>       INNER JOIN visits ON visits.url=urls.id
>   - name: UserRegex
>     default: .
>     
> export: |
>      LET browser_history(path,sql_query) =SELECT 
>             User, 
>             FullPath,
>             Mtime,
>             visit_url_id,
>             timestamp(winfiletime=url_visit_time * 10) as url_visit_time,
>             url_id,
>             visited_url,
>             url_title,
>             url_visit_count,
>             urls_typed_count,
>             timestamp(winfiletime=url_last_visit_time * 10) as url_last_visit_time
>         FROM sqlite(
>             file=path,
>             query=sql_query)
>           
>           
> sources:
>   - query: |
>       LET find_users = SELECT Uid, Name AS User, Directory
>         FROM Artifact.Windows.Sys.Users()
>         WHERE Name =~ UserRegex
>         
>       LET history_files = SELECT * FROM foreach(
>           row=find_users,
>           query={
>              SELECT User, FullPath, Mtime FROM glob(globs=Directory + KeyGlobs)
>           })
>           
>       SELECT * FROM foreach(row=history_files,
>             query={ SELECT * FROM browser_history(path=FullPath,sql_query=urlSQLQuery) 
>         })
> ```



> Good job!
> 
> I wonder if its worth consolidating to a general Browser history (its just really adding to the glob)?
> 
> (Credit the SQL below to one of my colleagues who also did similar join previously)
> 
> ```
> name: Windows.Applications.Browser.History
> description: |
>   Enumerate the users chrome and edge history. specifically the urls table and will join based on values from the visits table.
> 
> parameters:
>   - name: KeyGlobs
>     default: '\AppData\Local\{Google\Chrome,Microsoft\Edge}\User Data\*\History'
>   - name: urlSQLQuery
>     default: |
>       SELECT visits.url as visit_url_id, visits.visit_time as url_visit_time, urls.id as url_id, urls.url as visited_url, urls.title as url_title, urls.visit_count as url_visit_count,
>              urls.typed_count as urls_typed_count, urls.last_visit_time as url_last_visit_time
>       FROM urls
>       INNER JOIN visits ON visits.url=urls.id
>   - name: UserRegex
>     default: .
>     
> export: |
>      LET browser_history(path,sql_query) =SELECT 
>             User, 
>             FullPath,
>             Mtime,
>             visit_url_id,
>             timestamp(winfiletime=url_visit_time * 10) as url_visit_time,
>             url_id,
>             visited_url,
>             url_title,
>             url_visit_count,
>             urls_typed_count,
>             timestamp(winfiletime=url_last_visit_time * 10) as url_last_visit_time
>         FROM sqlite(
>             file=path,
>             query=sql_query)
>           
>           
> sources:
>   - query: |
>       LET find_users = SELECT Uid, Name AS User, Directory
>         FROM Artifact.Windows.Sys.Users()
>         WHERE Name =~ UserRegex
>         
>       LET history_files = SELECT * FROM foreach(
>           row=find_users,
>           query={
>              SELECT User, FullPath, Mtime FROM glob(globs=Directory + KeyGlobs)
>           })
>           
>       SELECT * FROM foreach(row=history_files,
>             query={ SELECT * FROM browser_history(path=FullPath,sql_query=urlSQLQuery) 
>         })
> ```

This is definitely a good way forward tbh. Consolidating the edge and chrome artefacts into a single one would be the best way forward IMO, so we don't get dupes.

**Comment by scudette** [comment]
Ideally long term we will have #1411 done and then we wont need all those little sqlite artifacts. The main issue is an explosion of small artifacts which are quite similar but just slightly different. It is better to try to handle them all as a single uber artifact.

**Comment by angry-bender** [comment]
> Ideally long term we will have #1411 done and then we wont need all those little sqlite artifacts. The main issue is an explosion of small artifacts which are quite similar but just slightly different. It is better to try to handle them all as a single uber artifact.

Yeah, my colleague just pointed this issue out. That will be better to have long term and simpler to manage too as the artefacts change. In the mean time, hopefully this pull will help for current instances until #1411 can be fixed.

**Comment by scudette** [comment]
Absolutely! Thank you for this PR 👍🏽

**Comment by angry-bender** [comment]
No problem :-)

**Pull request merged**